lib/srv/app/gcp/handler.go

```diff
@@ -0,0 +1,276 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcp
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"cloud.google.com/go/iam/credentials/apiv1/credentialspb"
+	"github.com/gravitational/oxy/forward"
+	oxyutils "github.com/gravitational/oxy/utils"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/api/utils/gcp"
+	"github.com/gravitational/teleport/lib/cloud"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/httplib"
+	"github.com/gravitational/teleport/lib/srv/app/common"
+	"github.com/gravitational/teleport/lib/utils"
+	awsutils "github.com/gravitational/teleport/lib/utils/aws"
+)
+
+// HandlerConfig is the configuration for an GCP app-access handler.
+type HandlerConfig struct {
+	// RoundTripper is the underlying transport given to an oxy Forwarder.
+	RoundTripper http.RoundTripper
+	// Log is the Logger.
+	Log logrus.FieldLogger
+	// Clock is used to override time in tests.
+	Clock clockwork.Clock
+
+	// CloudClients is a set of cloud clients that Teleport supports.
+	CloudClients cloud.Clients
+
+	// getAccessToken is a function for getting access token, pluggable for the sake of testing.
+	generateAccessToken generateAccessTokenFunc
+}
+
+// CheckAndSetDefaults validates the HandlerConfig.
+func (s *HandlerConfig) CheckAndSetDefaults() error {
+	if s.RoundTripper == nil {
+		tr, err := defaults.Transport()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		s.RoundTripper = tr
+	}
+	if s.Clock == nil {
+		s.Clock = clockwork.NewRealClock()
+	}
+	if s.Log == nil {
+		s.Log = logrus.WithField(trace.Component, "gcp:fwd")
+	}
+
+	if s.CloudClients == nil {
+		s.CloudClients = cloud.NewClients()
+	}
+
+	if s.generateAccessToken == nil {
+		s.generateAccessToken = s.generateAccessTokenDefaultImpl
+	}
+
+	return nil
+}
+
+type generateAccessTokenFunc func(ctx context.Context, serviceAccount string, scopes []string) (*credentialspb.GenerateAccessTokenResponse, error)
+
+func (s *HandlerConfig) generateAccessTokenDefaultImpl(ctx context.Context, serviceAccount string, scopes []string) (*credentialspb.GenerateAccessTokenResponse, error) {
+	client, err := s.CloudClients.GetGCPIAMClient(ctx)
```

Comment:
Don't we need close the client by defer client.Close after access token is generated ?

Reply:
We do. I've added the tests to ensure this is done too.

Reply:
@smallinsky Actually, turns out that we shouldn't close the client. If we do that it will cause errors when a next request comes, and a new token needs to be generated:
```
2023-01-05T11:00:44+01:00 DEBU [GCP:FWD]   Failed to process request. error:[
ERROR REPORT:
Original Error: *apierror.APIError rpc error: code = Canceled desc = grpc: the client connection is closing
Stack Trace:
	github.com/gravitational/teleport/lib/srv/app/gcp/handler.go:299 github.com/gravitational/teleport/lib/srv/app/gcp.(*handler).generateAccessToken
	github.com/gravitational/teleport/lib/srv/app/gcp/handler.go:267 github.com/gravitational/teleport/lib/srv/app/gcp.(*handler).getToken.func1.1
	github.com/gravitational/teleport/lib/utils/fncache.go:135 github.com/gravitational/teleport/lib/utils.FnCacheGet[...].func1
	github.com/gravitational/teleport/lib/utils/fncache.go:199 github.com/gravitational/teleport/lib/utils.(*FnCache).get.func1
	runtime/asm_arm64.s:1172 runtime.goexit
User Message: rpc error: code = Canceled desc = grpc: the client connection is closing] gcp/handler.go:183
```

Reply:
Ah, The `s.CloudClients.GetGCPIAMClient(ctx)` returns a global shared client so the client should stay unclosed so my comment is not invalid.